odl/discr/tensor_ops.py

```diff
@@ -0,0 +1,610 @@
+
+"""Operators defined for tensor fields."""
+
+# Imports for common Python 2/3 codebase
+from __future__ import print_function, division, absolute_import
+from future import standard_library
+standard_library.install_aliases()
+from builtins import super
+
+import numpy as np
+
+from odl.operator.operator import Operator
+from odl.set.sets import RealNumbers, ComplexNumbers
+from odl.space.fspace import FunctionSpace
+from odl.space.pspace import ProductSpace
+
+
+__all__ = ('PointwiseOperator', 'PointwiseNorm', 'PointwiseInner')
+
+
+class PointwiseOperator(Operator):
+
+    """Abstract operator for point-wise tensor field manipulations.
+
+    A point-wise operator acts on a space of vector or tensor fields,
+    i.e. a power space ``X^d`` of a discretized function space ``X``.
+    Its range is the power space ``X^k`` with a possibly different
+    number ``k`` of components.
+
+    For example, if ``X`` is a `DiscreteLp` space, then
+    ``ProductSpace(X, d)`` is a valid domain for any positive integer
+    ``d``.
+
+    Currently, only vector fields, i.e. one-dimensional products of
+    ``X``, are supported.
+
+    See also
+    --------
+    ProductSpace
+    """
+
+    def __init__(self, domain, range, linear=False):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        domain, range : {`ProductSpace`, `Discretization`}
+            Spaces of vector fields between which the operator maps.
+            They have to be either power spaces of the same base space
+            ``X`` or the base space itself (only one of them).
+        linear : `bool`, optional
+            If `True`, assume that the operator is linear.
+        """
+        if isinstance(domain, ProductSpace):
+            if not domain.is_power_space:
+                raise TypeError('domain {!r} is not a power space.'
+                                ''.format(domain))
+
+            if not isinstance(domain[0].uspace, FunctionSpace):
+                raise TypeError('domain base space {!r} is not a function '
+                                'space discretization.'.format(domain[0]))
+
+            dom_base = domain[0]
+        else:
+            dom_base = domain
+
+        if isinstance(range, ProductSpace):
+            if not range.is_power_space:
+                raise TypeError('range {!r} is not a power space.'
+                                ''.format(range))
+
+            if not isinstance(range[0].uspace, FunctionSpace):
+                raise TypeError('range base space {!r} is not a function '
+                                'space discretization.'.format(range[0]))
+
+            ran_base = range[0]
+        else:
+            ran_base = range
+
+        if dom_base != ran_base:
+            raise ValueError('domain and range have different base spaces '
+                             '({!r} != {!r}).'
+                             ''.format(domain[0], range[0]))
```

Comment:
This will fail if the space is not a productspace, should be `dom_base, ran_base`